test/stream_router_test.exs

```diff
@@ -0,0 +1,113 @@
+defmodule TestSink do
+  defstruct pid: nil, forward_to: nil, val: nil
+  use StreamRouter
+
+  def start_link(args, opts \\ []) do
+    GenServer.start_link(__MODULE__, args, opts)
+  end
+
+  def init({val, forward}) do
+    {:ok, %__MODULE__{val: val, forward_to: forward}}
+  end
+
+  def handle_info({:stream_up, source}, state) do
+    StreamRouter.ask_async(source, 1)
+    {:noreply, state}
+  end
+
+  def handle_event({:eos, _}, _, %{forward_to: forward} = state) do
+    send forward, :eos
+    {:noreply, state}
+  end
+
+  def handle_event(new_val, source, state) do
+    StreamRouter.ask_async(source, 1)
```

Comment:
Note: ask here at the moment cannot be sync as this results in a race condition. The StreamRouter process is already waiting on a response from the sink process as `$gen_route` is synchronous, and thus can't respond to a synchronous `ask` message if it's done within the handle_event callback.